src/DDTrace/Integrations/Kafka/KafkaIntegration.php

```diff
@@ -0,0 +1,233 @@
+        $span = $hook->span();
+        KafkaIntegration::setupCommonSpanMetadata($span, Tag::KAFKA_PRODUCE, Tag::SPAN_KIND_VALUE_PRODUCER, Tag::MQ_OPERATION_SEND);
+
+        $span->meta[Tag::MQ_DESTINATION] = $producerTopic->getName();
+        $span->meta[Tag::MQ_DESTINATION_KIND] = Type::QUEUE;
+
+        $conf = ObjectKVStore::get($producerTopic, 'conf');
+        KafkaIntegration::addProducerSpanMetadata($span, $conf, $hook->args);
+
+        $headers = \DDTrace\generate_distributed_tracing_headers();
+        $hook->args = $this->injectHeadersIntoArgs($hook->args, $headers);
+        $hook->overrideArguments($hook->args);
+    }
+
+    public static function addProducerSpanMetadata($span, $conf, $args)
+    {
+        self::addMetadataToSpan($span, $conf);
+        $span->metrics[Tag::KAFKA_PARTITION] = $args[0];
+        $span->metrics[Tag::MQ_MESSAGE_PAYLOAD_SIZE] = strlen($args[2]);
+        if (isset($args[3])) {
+            $span->meta[Tag::KAFKA_MESSAGE_KEY] = $args[3];
+        }
+    }
+
+    private function injectHeadersIntoArgs(array $args, array $headers): array
+    {
+        // public RdKafka\ProducerTopic::producev (
+        //      integer $partition ,
+        //      integer $msgflags ,
+        //      string $payload [,
+        //      string $key = NULL [,
+        //      array $headers = NULL [,
+        //      integer $timestamp_ms = NULL [,
+        //      string $opaque = NULL
+        // ]]]] ) : void
+        $argsCount = count($args);
+        if ($argsCount >= 5) {
+            $args[4] = array_merge($args[4] ?? [], $headers);
+        } elseif ($argsCount === 4) {
+            $args[] = $headers;
+        } else { // $argsCount === 3
+            $args[] = null;  // $key
+            $args[] = $headers;
+        }
+        return $args;
+    }
+
+    private function installConsumerHooks()
+    {
+        $integration = $this;
+
+        $consumerMethods = [
+            'RdKafka\KafkaConsumer::consume',
+            'RdKafka\Queue::consume'
+        ];
+
+        foreach ($consumerMethods as $method) {
+            \DDTrace\install_hook(
+                $method,
+                function (HookData $hook) use ($integration) {
+                    $hook->data['start'] = \DDTrace\now();
+                },
+                function (HookData $hook) use ($integration) {
+                    $integration->processConsumedMessage($hook);
+                    $integration->setupKafkaConsumeSpan($hook, $this);
+                    \DDTrace\close_span();
+                }
+            );
+        }
+    }
+
+    public function processConsumedMessage(HookData $hook)
+    {
+        /** @var \RdKafka\Message $message */
+        $message = $hook->returned;
+
+        if ($message) {
+            $headers = array_filter(KafkaIntegration::extractMessageHeaders($message->headers ?? []), function($value) {
+                return $value !== null;
+            });
+
+            if (!empty($headers)) {
```

Comment:
```suggestion
            if ($message->headers && $link = SpanLink::fromHeaders($message->headers)) {
```
`SpanLink::fromHeaders()` returns null if there's nothing to extract or extraction failed. This is better than checking for specific headers existence (and you also forgot b3).

(you may then use `$link` directly in the `$span->links[] =` assignment too.